pkg/package-server/provider/registry.go

```diff
@@ -131,43 +157,176 @@ func (p *RegistryProvider) syncCatalogSource(obj interface{}) (syncError error)
 		return
 	}
 
-	key := sourceKey{source.GetName(), source.GetNamespace()}
-	client, ok := p.getClient(key)
-	if ok && source.Status.RegistryServiceStatus.ServiceName != "" {
-		logger.Info("update detected, attempting to reset grpc connection")
-		client.conn.ResetConnectBackoff()
+	address := source.Address()
+	logger = logger.WithField("address", address)
+
+	key := resolver.CatalogKey{
+		Namespace: source.GetNamespace(),
+		Name:      source.GetName(),
+	}
+	if sourceMeta := p.sources.GetMeta(key); sourceMeta != nil && sourceMeta.Address == address {
+		// If the address hasn't changed, don't bother creating a new source
+		logger.Debug("catalog address unchanged, skipping source creation")
+		return
+	}
+
+	logger.Info("connecting to source")
+	if _, syncError = p.sources.Add(key, address); syncError != nil {
+		logger.Warn("failed to create a new source")
+	}
 
-		ctx, cancel := context.WithTimeout(context.TODO(), defaultConnectionTimeout)
-		defer cancel()
+	return
+}
 
-		changed := client.conn.WaitForStateChange(ctx, connectivity.TransientFailure)
-		if !changed {
-			logger.Debugf("grpc connection reset timeout")
-			syncError = fmt.Errorf("grpc connection reset timeout")
-			return
+func (p *RegistryProvider) syncSourceState(state registrygrpc.SourceState) {
+	key := state.Key
+	logger := logrus.WithFields(logrus.Fields{
+		"action": "sync source",
+		"source": key,
+		"state":  state.State,
+	})
+	logger.Debug("source state changed")
+
+	timeout, cancel := context.WithTimeout(context.Background(), cacheTimeout)
+	defer cancel()
+
+	var err error
+	switch state.State {
+	case connectivity.Ready:
```

Comment:
It’s unlikely, but there are cases where you might not receive an event for this (for example, if the connection is already in the ready state before it’s events are watched, or if the states change too rapidly). In practice this doesn’t seem to be an issue for grpc connections, but I wanted to mention it. 

(This is due to relying on the grpc wait for state change, which requires you specify the state you’re waiting to change away from)

It’s not an issue in catalog-operator because the state change is used as an optimization and requeues a catalogsource for processing - if an event is missed like that, a periodic resync will fix it.

Reply:
Fair -- We can make caching more aggressive in the sync function if you want.